### ISSUE TITLE
 Items in the Cape slot from trinkets are not swapped back

### ISSUE THREAD

**Issue by el0c** [Szum123321/elytra_swap#11]
The trinkets compatibility is working, however it doesn't swap the previous item back in to the cape slot like the mod swaps the chest plate back to the chest slot when trinkets isn't installed. I was wondering if this functionality could be added to the trinkets cape swapping as well?

Playing on the 1.15.2 1. 50 version.

**Comment by Szum123321** [comment]
Sadly Eltrya Swap searches for the first elytra or chest plate in the inventory. This mechanism would not really work with other trinkets, so for now now I think that it should stay as it is. Maybe later I will add some sort of enchantment that would allow some items to be fawored.

**Comment by el0c** [comment]
So looking over the source, (with my admittedly limited java knowledge) it looks to me like this method already creates a number of issues.
What if a player is looting the end where other chest plates might be plentiful? Same problem with an elytra. Wouldn’t it be possible it uses a fresh one you found instead of the mending enchanted one you’re using to get around?

From what I can tell and how you worded it, it seems like it’s just going to choose the first one it finds in the list created by your FlatInventory code.

Hmmm.... The enchantment idea sounds like it might be more difficult than just updating how it searches for chest plates/capes. To both the end user and yourself, in my opinion.

Users would have to spend EXP to get the enchantment and the code would have to look at NBT data of all inventory items anyways; so wouldn’t it be simpler to store the slot you switched the elytra from and simply move the item to that slot, then swap back from that slot upon landing? The only issue I see with that is if the player moves the item while flying. You could compensate for (that)/only store and search for the item by all of its NBT flags. Not sure how you can differentiate between one basic piece of armor and the next if neither are named, have damage or any enchantments though....
I guess you could add your own NBT flag on swap with the elytra that basically “isSwapItem” and remove the flag when it’s moved back to the slot?

This is all pseudo code from an armchair developer though; maybe you could tell me why that wouldn’t be possible in this case?

**Comment by Szum123321** [comment]
Yes, you are right. Current method creates quite a few issues but when first writing this mod I thought that those problems which you have listed above weren't that important. I would like to notice that now Elytra Swap selects items form the lowest slots, do it is possible to just have your chestplate/elytra in the first slot. That way no matter what, unless you also have inventory sorter of some sort, player can select which items would be used. 

Now, your first suggestion sounds like a good idea, but really it has the same issues. The most important one being that FlatInventory indexes are dependent on current inventory arrangement (shulker boxes) therefore this approach also won't be resistant to movements in the inventory. Your second proposition is much more interesting as it mitigates the main problem. 

Yet I still think that the best way to do this would be to have priority list of some sort and put all the found items that can go into given slot and then select the item with the highest 'value'. Value of an item could be determined for example this way:

- 1 for the first item to be found
- 2 for the item with "isSwapItem” tag
- 3 for item with level 1 enchantment (something like swappiness....).
 - 4 for item with level 2 enchantment 
 - and so on.

This technique would not only get rid of the problems we talked above, but also would give the player chance to select which exact item would be used and will not require players to spend their precious xp.

So what do you think about this?

**Comment by el0c** [comment]
I think that's actually a really good system. The "isSwapItem" provides compatibility with other trinkets in the cape slot and weird items like backpacks that might be in the chestplate slot;
the optional enchant and enchant levels sounds neat! I like the versatility it would provide.

Also, not sure what else to call the enchantment but swapiness hahaha! Gave it thought for a while but there's already a mod called Soulbound and Soul Bond was the only thing I came up with but that seems too close.

**Comment by Szum123321** [comment]
Great! So when I find some time I will do it.
